File: beginner_tutorials/scripts/key_reader.cpp
#include "ros/ros.h"
#include "geometry_msgs/Twist.h"

int main(int argc, char **argv){
    ros::init(argc, argv, "control_car");
    ros::NodeHandle node_obj;
    ros::Publisher control_car_publisher=node_obj.advertise<geometry_msgs::Twist>("/cmd_vel", 10);
    ros::Rate loop_rate(10);
    geometry_msgs::Twist msg;
    // msg.linear.x=0;
    printf("Use w,a,s,d to move the bot and press q for exit.\n");
    std::string inputString;
    while(ros::ok()){
        std::getline(std::cin, inputString);
        if(inputString.compare("w") == 0){
        msg.linear.x=msg.linear.x+0.01;
        }
        if(inputString.compare("q") == 0){
        msg.linear.x=0;
        }
        if(inputString.compare("s") == 0){
        msg.linear.x=msg.linear.x-0.01;
        }
        ROS_INFO("Publishing\n");
        control_car_publisher.publish(msg);
        loop_rate.sleep();
    }
    
    return 0;
}

File: beginner_tutorials/scripts/key_reader.py
#!/usr/bin/env python3

import sys
import tty
import termios
import time
import rospy
#ros msgs
from geometry_msgs.msg import Twist
# from std_msgs.msg import String


def getTerminalSettings():
    if sys.platform == 'win32':
        return None
    return termios.tcgetattr(sys.stdin)


def readKey(settings):
    if sys.platform == 'win32':
        # getwch() returns a string on Windows
        key = msvcrt.getwch()
    else:
        tty.setraw(sys.stdin.fileno())
        # sys.stdin.read() returns a string on Linux
        key = sys.stdin.read(1)
        termios.tcsetattr(sys.stdin, termios.TCSADRAIN, settings)
    return key

rospy.init_node('rc_control',anonymous=True)

cmd_vel= rospy.Publisher('/cmd_vel',Twist,queue_size=20)
# rc_dir_pub= rospy.Publisher('/dir_selection',Int32,queue_size=20)

if __name__ == '__main__':

    settings=getTerminalSettings()

    try:
        
        r = rospy.Rate(10)
        key='\0'
        twist_msg = Twist()

        twist_msg.linear.x = 0.0
        cmd_vel.publish(twist_msg)
        print("Velocity published: linear.x = ", twist_msg.linear.x)


        while True:
        
            key=readKey(settings)
            
            if key == 'w':
                twist_msg.linear.x = twist_msg.linear.x+0.01
                print("Velocity published: linear.x = ",twist_msg.linear.x)
            elif key == 's':
                twist_msg.linear.x = twist_msg.linear.x+-0.01
                print("Velocity published: linear.x = ",twist_msg.linear.x)
            else:
                
                print("velocity published is null")
            
            if key == 'd':
                twist_msg.angular.z = twist_msg.angular.z+0.01
                print("Changing direction : ",twist_msg.angular.z)
            elif key == 'a':
                twist_msg.angular.z = twist_msg.angular.z+-0.01
                print("Changing direction : ",twist_msg.angular.z)
            else:
                
                print("No change in direction published")
            cmd_vel.publish(twist_msg)
            if key=='p':
                break
            
            r.sleep()
            
        print("controller terminated")
                
            
            
    except KeyboardInterrupt:
        print("interupted!!!")

    except Exception as e:
        print(e)